feature `dso mv`

tschwarzl opened the pull request:
Closes #132

implemtend `dso mv` which moves and or renames stages or folder and updates references to it in dvc.yaml, params.in.yaml and non-hidden src files. When mv a stage, README.md is also updated. A minimal unit test is added. Minimal documentation fo rnow. 

-   [x] CHANGELOG.md updated
-   [x] Added minimal tests (For bug fixes or new features)
-   [x] Documentation updated (if applicable)

Commit:
graceful exit when exiting template selection

Commit:
added mv functionality

Commit:
changed so that all files in src are renamed

Commit:
added all src files for stage update

tschwarzl committed:
Merge branch 'main' into feature/rename-stage

tschwarzl committed:
Update CHANGELOG.md

Commit:
added command reference for mv

grst commented on src/dso/cli/_mv.py:
```diff
@@ -0,0 +1,175 @@
+
+    project_root = get_project_root(Path(getcwd()))
+
+    source_relative_path = get_direct_path_from_root(project_root, source)
+    target_relative_path = get_direct_path_from_root(project_root, target)
```
Can `get_direct_path_from_root` be replaced with pathlib's `source.relative_to(project_root)`? 

In that case the `normpath()` call would also be superflous I think.

grst reviewed (commented):
First round of review, more to come:

1) In terms of structure, please try to keep the CLI definition and the program logic separate. For all the other functions, I keep the logic in `/src/_xxx.py`, while the CLI is a single function added to `/src/cli/__init__.py`. This is done for two reasons:
  * it makes the CLI more responsive, since not all 3rd party packages get imported immediately, but only when they are needed
   * separation of concerns. The functions in `/src/_xx.py` can also be called from somewhere else in the code, while the CLI is user-facing. 

2) Use pathlib. Python has both the `os.path` module and `pathlib`. You mix both of them which leads to a lot of conversions. Pathlib is more modern, concise and object-oriented and for new projects, there's basically no reason to use any `os.path` function anymore. Using `pathlib` consistently will get the line count quite a bit here I think.

tschwarzl commented on CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@ and this project adheres to [Semantic Versioning][].
 ### Fixes
 
 -  Exit gracefully when aborting template selections.
+-  added `dso mv` for moving and renaming stages or folders while updating names and references
```
Thank you for pointing this out.

tschwarzl committed:
use of Pathlib

Co-authored-by: Gregor Sturm <gregor.sturm@boehringer-ingelheim.com>

tschwarzl committed:
pre-commit autofixes

tschwarzl commented on src/dso/cli/_mv.py:
```diff
@@ -0,0 +1,175 @@
+
+
+@click.command(name="mv")
+@click.argument("source", type=click.Path(exists=True, file_okay=True, dir_okay=True, path_type=Path))
+@click.argument("target", type=click.Path(file_okay=True, dir_okay=True, path_type=Path))
+def dso_mv(source: str | None = None, target: str | None = None):
+    """Move and rename a stage or a folder and update references to it
+
+    A stage or folder is renamed with references to it. In a stage, the dvc.yaml, params.in.yaml
+    and src files. In other folder or stages, dvc.yaml, params.in.yaml and src files are updated.
+    """
+    # rewrite to relative path roots
+    if source is None:
+        source = Prompt.ask("[bold]Please enter the name of the stage or folder to be renamed")
+    if target is None:
+        target = Prompt.ask("[bold]Please enter the new stage or folder name")
```
Thank you for the comment. I will remove the rich Prompt, which is a left-over from the older PR

tschwarzl replied:
Suggestion were implemented

tschwarzl committed:
removed casting

Co-authored-by: Gregor Sturm <gregor.sturm@boehringer-ingelheim.com>

tschwarzl committed:
removed casting

Co-authored-by: Gregor Sturm <gregor.sturm@boehringer-ingelheim.com>

tschwarzl committed:
pre-commit autofixes

grst commented on src/dso/cli/_mv.py:
```diff
@@ -0,0 +1,175 @@
+
+def update_references_in_file(file: Path, pattern: str, replacement: str):
+    """Update all pattern in file"""
+    if os.path.exists(file):
```
Is it intended to silently do nothing if the file does not exist?

grst commented on src/dso/cli/_mv.py:
```diff
@@ -0,0 +1,175 @@
+    update_files_in_src(src_path, source_base, target_base, str(source_relative_path), str(target_relative_path))
+
+    if os.path.dirname(source_relative_path) != os.path.dirname(target_relative_path):
+        log.info(
```
```suggestion
        log.warning(
```

grst commented on src/dso/cli/_mv.py:
```diff
@@ -0,0 +1,175 @@
+            for file in files:
+                old_file_path = os.path.join(root, file)
+                if not file.startswith(".") and os.path.isfile(old_file_path):
+                    new_file_name = file.replace(source_base, target_base)
```
I think string-based maniuplations of dirnames are dangerous, I'd rather again use `Path.relative_to`

tschwarzl replied:
suggestion was implemented

grst commented on src/dso/cli/_mv.py:
```diff
@@ -0,0 +1,175 @@
+        log.debug(f"Updating {file}: replacing '{pattern}' with '{replacement}'")
+        try:
+            content = Path(file).read_text()
+            updated_content = content.replace(pattern, replacement)
```
There will be edge cases, e.g. stage names like `preprocessing` could also be used as a word in the text, or there could be references to another stage, e.g. "rnaseq_preprocessing". 

But I don't think there's a lot we can do about it. Just try to address more common edge cases as users discover them.

grst commented on src/dso/cli/_mv.py:
```diff
@@ -0,0 +1,175 @@
+from dso._util import get_project_root
+
+
+def update_all_references_to_source(project_root: Path, source_absolute_path: str, target_absolute_path: str):
+    """In the whole project, search for relative references to the old location and update to relative reference to the new location."""
+    log.debug("[yellow]Updating all references to source")
+    log.debug(f"{project_root} - {source_absolute_path} - {target_absolute_path}")
+    for dvc_dir in project_root.rglob("*"):
+        if dvc_dir.is_dir() and (dvc_dir / "dvc.yaml").exists() and dvc_dir != target_absolute_path:
+            source_direct_path = os.path.relpath(source_absolute_path, start=dvc_dir)
+            target_direct_path = os.path.relpath(target_absolute_path, start=dvc_dir)
+
+            log.info(f"Relatives path from {dvc_dir}: {source_direct_path} to {target_direct_path}")
+
+            update_references_in_file(dvc_dir / "dvc.yaml", source_direct_path, target_direct_path)
+            update_references_in_file(dvc_dir / "params.in.yaml", source_direct_path, target_direct_path)
+
+            # Iterate over all non-hidden files in the src subdirectory of dvc_dir
+            src_subdir = dvc_dir / "src"
+            if src_subdir.exists() and src_subdir.is_dir():
+                for file_path in src_subdir.rglob("*"):
+                    if file_path.is_file() and not file_path.name.startswith("."):
+                        update_references_in_file(file_path, source_direct_path, target_direct_path)
```
Don't you need to somehow call this recursively? As far as I understand this iterates over all directories in the project root and works on those that have a `dvc.yaml`. But what if you have stages nested in subdirectories, e.g. 

```
.
├── RNAseq
│   ├── dvc.yaml
│   └── 01_preprocessing
│       └── dvc.yaml
└── TSO500
    ├── dvc.yaml
    ├── 01_preprocessing
    │   └── dvc.yaml
    └── 02_visualization
        └── dvc.yaml
```

tschwarzl replied:
Was implemented recursively with fixes

tschwarzl committed:
Update src/dso/cli/_mv.py

Co-authored-by: Gregor Sturm <gregor.sturm@boehringer-ingelheim.com>

Commit:
merge from main

Commit:
changed most to pathlib

Commit:
added error message when file does not exist

Commit:
error when src dir does not exist

Commit:
rewrote to pathlib

Commit:
renamed function

Commit:
issue with fixure dso_project_with_multiple_stages

tschwarzl committed:
pre-commit autofixes

grst committed:
Merge branch 'main' into feature/rename-stage

grst committed:
Ensure dso create can handle absolute paths as stage name

grst committed:
Use stage name when instantiating template, not path

grst committed:
Make dso create work with paths

grst committed:
Merge branch 'main' into feature/rename-stage

Commit:
added unit tests for multi level stages project and recursive updating of path. fixed issues of correct path recognition

tschwarzl committed:
pre-commit autofixes

Commit:
organised imports

Commit:
Merge branch 'feature/rename-stage' of https://github.com/Boehringer-Ingelheim/dso into feature/rename-stage

Commit:
moved mv command to common dso interface

Commit:
update description

Commit:
updated changelog and anntoation

Commit:
update doc

Commit:
updated cli command reference for mv

tschwarzl commented:
Closes #132 #136.

Implemented suggestions and improvements to make the `dso mv` command available as experimental feature. 

- [x] add feature request: updating outbound relative paths for moved stages or folders [#159]
- [x] Implemented proper use of Pathlib
- [x] Implemented proper recursive update 
- [x] BugFix: Only update references if there are no '/' or '\' in front of path  
- [x] Added unit test for complex project setup testing.
- [x] updated Changelog
- [x] updated CLI references

Commit:
minor update to CHANGELOG

Commit:
added parametrize unit tests

tschwarzl commented:
- [x] Modified unit test for moving stages in project with dependencies to use `pytest.mark.parametrize`

tschwarzl commented:
Todos:

- [x] Add a basic unit tests for renaming folders

Commit:
added unit tests for mv folder

Commit:
added unit tests

tschwarzl committed:
pre-commit autofixes

Commit:
added fail unit tests

Commit:
merge branch 'feature/rename-stage' of https://github.com/Boehringer-Ingelheim/dso into feature/rename-stage

Commit:
updated changelog for dso mv --increment-prefix

tschwarzl commented:
I have added the increment feature, since it added minor changes to `dso mv` PR. Comprehensive unit tests were added.

grst commented on src/dso/cli/_create.py:
```diff
@@ -51,24 +54,32 @@
-    target_dir = Path(getcwd()) / params["name"]
+    target_dir = Path.cwd() / params["name"]
+    # ensure `name` is only the stage name even when path is specified
+    params["name"] = Path(params["name"]).name
```
@grst: note to self, double-check, I think I played with this and it made some issues somewhere in some template

grst replied:
this seems to work as expected. Not sure what the issue was back then.

grst commented on src/dso/_mv.py:
```diff
@@ -0,0 +1,308 @@
+    and need to be adjusted manually.
+    """
+    log.info(f"[yellow]Renaming '{source}' to '{target}'.")
+    log.info("[red]Warning: This feature is experimental.")
```
```suggestion
    log.info("[red]Warning: This feature is experimental. Behavior may change in the future. Please report issues on GitHub.")
```

grst commented on src/dso/_mv.py:
```diff
@@ -0,0 +1,308 @@
+        target_absolute_path,
+    )
+
+    update_references_to_source_recursively(project_root, source_absolute_path, target_absolute_path)
```
Why pass absolute paths here, only to convert them again to relative paths first thing in the function?

tschwarzl replied:
This is not a simple conversion; it needs to determine relative paths between nested structures. See longer explanation in response to main comment below.

grst replied:
ok, fair enough. I still feel a lot of redundant information is passed around as function parameters, but ultimately it's not that important.

grst commented on src/dso/_mv.py:
```diff
@@ -0,0 +1,308 @@
+
+    source_relative_path = source.relative_to(project_root)
+    target_relative_path = target.relative_to(project_root)
+
+    target_relative_dir = target_relative_path.parent
+
+    log.debug(f"Relative source path: {source_relative_path}")
+    log.debug(f"Relative target path: {target_relative_path}")
+
+    source_absolute_path = project_root / source_relative_path
+    target_absolute_path = project_root / target_relative_path
+    target_absolute_dir = project_root / target_relative_dir
+
+    source_base = source_absolute_path.name
+    target_base = target_absolute_path.name
+
+    log.debug(f"Base source: {source_base}")
+    log.debug(f"Base target: {target_base}")
+    log.debug(f"Absolute source path: {source_absolute_path}")
+    log.debug(f"Absolute target path: {target_absolute_path}")
+    log.debug(f"Absolute target base directory: {target_absolute_dir}")
+
+    if not source_absolute_path.exists():
+        log.error(f"[red]{source} does not exist!")
+        exit(1)
+
+    # Currently is not allowed to move the folder or stage to a directory
+    # which does not exist.
+    if not target_absolute_dir.exists():
+        log.error(f"[red]{target_relative_dir} does not exist. Target base directory must already exist.")
+        exit(1)
+
+    if target_absolute_path.exists():
+        log.error(f"[red]{target} already exists!")
+        exit(1)
+
+    update_source(
+        source_base,
+        target_base,
+        source_relative_path,
```
I think in this whole section, we could get rid of the absolute paths and instead keep the relative paths and pass also the project_root around when necessary. 

Overall, I think there are superflous conversion between absolute and relative paths which makes it harder to reason about the code.

tschwarzl replied:
I would like to keep the option that the function can handle absolute paths. i can split this function up in to two, one which takes relative paths

-- edit --: see main comment below. 
the main function does the conversions and passes arguments down, the other function need to determine relative paths between nested structures.

grst commented on src/dso/_mv.py:
```diff
@@ -0,0 +1,308 @@
+    # split source into two parts, first with the same length of prefix
+    # second part is the rest
+    source_prefix = source.name[: len(target_prefix)]
+    # source_base = source.name[len(target_prefix) :]
```
source_base doesn't seem to be used, comment can be cleaned up.

grst commented on src/dso/_mv.py:
```diff
@@ -0,0 +1,308 @@
+    # source_base = source.name[len(target_prefix) :]
+
+    # determine how many digits are at the end of new_prefix starting from the last character
+    target_prefix_number_of_digits = 0
+    for c in reversed(target_prefix):
+        if c.isdigit():
+            target_prefix_number_of_digits += 1
+        else:
+            break
+    if target_prefix_number_of_digits == 0:
+        log.error(f"[red]Prefix '{target_prefix}' must end with at least one digit.")
+        exit(1)
+
+    # determine how many digits are at the end of source.name starting from the last character
+    source_prefix_number_of_digits = 0
+    for c in reversed(source_prefix):
+        if c.isdigit():
+            source_prefix_number_of_digits += 1
+        else:
+            break
```
I suggest making this a helper function, e.g. 

```python
def _get_n_numeric(prefix) -> int:
    for i, c in enumerate(reversed(prefix)):
        if not c.isdigit():
             return i
    return len(prefix)
```

tschwarzl replied:
will implement it

grst commented on src/dso/_mv.py:
```diff
@@ -0,0 +1,308 @@
+        exit(1)
+
+    # get the numeric indeces and calculate the offset
+    source_prefix_index = int(source_prefix[len(source_prefix) - source_prefix_number_of_digits :])
```
Python supports indexing from behind

```suggestion
    source_prefix_index = int(source_prefix[-source_prefix_number_of_digits :])
```

grst commented on src/dso/_mv.py:
```diff
@@ -0,0 +1,308 @@
+    offset = target_prefix_index - source_prefix_index
+
+    # get the stem of the prefixes
+    source_prefix_stem = source_prefix[: len(source_prefix) - source_prefix_number_of_digits]
```
```suggestion
    source_prefix_stem = source_prefix[: - source_prefix_number_of_digits]
```

grst reviewed (commented):
Looks good! 
I have a few more minor requests! Also I still want to give it a try myself before merging.

tschwarzl committed:
Update src/dso/_mv.py

Co-authored-by: Gregor Sturm <gregor.sturm@boehringer-ingelheim.com>

tschwarzl committed:
pre-commit autofixes

tschwarzl commented:
> Looks good! I have a few more minor requests! Also, I still want to give it a try myself before merging.

Thank you for the review and the python tips.

Regarding the path conversions in `update_references_to_source_recursively`, I would like to clear things up: this function needs to determine relative paths within nested structures: e.g. how to get from `0100_A/0102` to `0100_B/0101/` as used in `params.in.yaml` (`../0101/`). Passing relative paths will not circumvent the issue of determining relative paths to other (nested) directories. 

Therefore, I have decided to determing this from absolut paths - which technically could also be relative paths, but I do not see benefits.

In other functions, like `update_files_in_src_folder`, or `update_source` - the file paths, relative and absolute were determined in the main function `_mv` and are passed down.

I would argue for leaving this once you are happy with testing.

Commit:
implemented feedback from review

grst commented on src/dso/_mv.py:
```diff
@@ -0,0 +1,308 @@
+            updated_content = sub(rf"(?<![\\/]){escape(pattern)}", replacement, content)
+            file.write_text(updated_content)
+            if content != updated_content:
+                log.info(f"Updated references in {file}: {content} -> {updated_content}")
```
This is quite verbose, especially for longer files.

<img width="1403" height="430" alt="image" src="https://github.com/user-attachments/assets/0f42d9d1-6ebb-44b1-bdd1-c7022197a200" />

Maybe we can switch to a diff-like output here? 



There's a stdlib function for that: [difflib.unified_diff](https://docs.python.org/3/library/difflib.html#difflib.unified_diff)

grst commented on src/dso/_mv.py:
```diff
@@ -0,0 +1,308 @@
+    target_relative: Path,
+):
+    """Rename all files which contain stage name, for all update the stage"""
+    if path.exists():
+        for file_path in path.rglob("*"):
+            if file_path.is_file() and not file_path.name.startswith("."):
+                new_file_name = file_path.name.replace(str(source_base), str(target_base))
+                new_file_path = file_path.parent / new_file_name
+                log.debug(f"Renaming file {file_path} to {new_file_path}")
+                try:
+                    file_path.rename(new_file_path)
+                except OSError as e:
+                    log.error(f"[red]Failed to rename {file_path}: {e}")
+
+                update_references_in_file(new_file_path, str(source_relative), str(target_relative))
```
Maybe we should constrain this to files tracked by git? 
I just tested on a case where was a `.quarto` folder with quarto chache in the src directory. It contained a lot of binary files on which the rename failed. It took forever and it caused a ton of error log messages. 

There's a function `git_list_files` in `_util.py` that can be used.

grst commented on src/dso/_mv.py:
```diff
@@ -0,0 +1,308 @@
+        update_references_in_file(current_path / filename, source_direct_path, target_direct_path)
+
+    # Iterate over all non-hidden files in the src subdirectory of dvc_dir
+    src_subdir = current_path / "src"
+    if src_subdir.exists() and src_subdir.is_dir():
+        for file_path in src_subdir.rglob("*"):
+            if file_path.is_file() and not file_path.name.startswith("."):
+                update_references_in_file(file_path, source_direct_path, target_direct_path)
+
+    for dvc_dir in current_path.rglob("*"):
+        # search all non-hidden folders with dvc.yaml inside which are not part of target
+        if (
+            dvc_dir.is_dir()
+            and not dvc_dir.name.startswith(".")
+            and (dvc_dir / "dvc.yaml").exists()
+            and dvc_dir != target_absolute_path
+        ):
+            update_references_to_source_recursively(dvc_dir, source_absolute_path, target_absolute_path)
```
Maybe we should constrain this to files tracked by git? 
I just tested on a case where was a `.quarto` folder with quarto chache in the src directory. It contained a lot of binary files on which the rename failed. It took forever and it caused a ton of error log messages. 

There's a function `git_list_files` in `_util.py` that can be used. 

<img width="1277" height="375" alt="image" src="https://github.com/user-attachments/assets/2fd8a183-0b4e-4c1d-a1de-699a4ac3c824" />

grst replied:
in another case I had a zarr folder (100k files in a folder) lying in one of the stages. `mv` got stuck on that, because it will read each and every file. 

The folder is gitignored, so it would have been solved by limiting to git files as well.

grst commented on src/dso/_mv.py:
```diff
@@ -0,0 +1,308 @@
+
+    if target_absolute_path.exists():
+        log.error(f"[red]{target} already exists!")
+        exit(1)
```
To be consistent with the mv command, the following would have to work

```
dso mv subfolder/mystage anotherfolder_that_exists
```

should move `subfolder/mystage` to `anotherfolder_that_exists/mystage`. Is this something we would like to support?

grst commented on src/dso/cli/__init__.py:
```diff
@@ -266,13 +267,56 @@ def command_wrapper(args):
+    required=False,
+)
+@click.option(
+    "--increment-prefix",
```
```suggestion
    "--increment-prefix",
    "-i",
```

increment-prefix is a pain to type, let's add an additional shorthand option

grst commented on src/dso/_mv.py:
```diff
@@ -0,0 +1,308 @@
+    for item in source_target_list:
+        log.info(f"[yellow]Renaming {item['source_path'].name} to {item['target_path'].name} in {parent_dir}")
+
+        if (item["target_path"] / "dvc.yaml").exists():
```
```suggestion
        if (item["source_path"] / "dvc.yaml").exists():
```

This should be source_path, no? 

When I tried `mv --increment-prefix`, it didn't update the stages, just moved them, probably because of that.

grst reviewed (changes requested):
In terms of code structure, I'm happy now, but I found a couple of practical issues while trying it out on a real example. See my comments below (and above).

grst committed:
Merge branch 'main' into feature/rename-stage

grst committed:
Address issues from code review

grst committed:
Merge remote-tracking branch 'origin/main' into feature/rename-stage

grst committed:
Do not check binary files

grst committed:
Avoid double-recursion
